cmd/vcfToFa/vcfToFa.go

```diff
@@ -11,31 +11,76 @@ import (
 	"log"
 )
 
-func vcfToFa(inVcfFilename string, inFaFilename string, outFaFilename string, useAlt bool) {
+func vcfToFa(inVcfFilename string, inFaFilename string, outFaFilename string, useAlt bool, multiFaMode bool, multiFaChromName string) {
 	var seqsOrdered []fasta.Fasta
 	var seqsLookup fasta.FastaMap
 	var vcfRecords <-chan vcf.Vcf
+	var firstTime bool = true
+	var prevPos int              //position of the previous variant. Used to enforce sorted order in multiFaMode.
+	var multiFaAppendName string //name of the appended record in multiFaMode.
+	var currAlnPos int
+	var outIndex int
+	var prevRefPos, prevAlnPos int = 0, 0
 
 	// We hang on to the ordered version of the fasta sequences so that we could
 	// output the edited sequences (those in the map) in the same order that
 	// we originally read them, but we do not have that currently
 	seqsOrdered = fasta.Read(inFaFilename)
+	if multiFaMode && useAlt {
+		var refSeqCopy []dna.Base
+		if len(seqsOrdered) == 0 {
+			log.Fatalf("Input fasta must contain at least one record in multiFaMode.")
+		}
+		refSeqCopy = make([]dna.Base, len(seqsOrdered[0].Seq))
+		copy(refSeqCopy, seqsOrdered[0].Seq)
+		multiFaAppendName = seqsOrdered[0].Name + "alt"
+		seqsOrdered = append(seqsOrdered, fasta.Fasta{Name: multiFaAppendName, Seq: refSeqCopy})
+		outIndex = len(seqsOrdered) - 1 //index of the output sequence.
+	}
 	seqsLookup = fasta.ToMap(seqsOrdered)
 
 	// we could do some sort of header check here to make sure the fa matches the vcf
 	vcfRecords, _ = vcf.GoReadToChan(inVcfFilename)
+
 	for v := range vcfRecords {
+		//first a block of code to enforce sorted order in multiFaMode.
+		if firstTime && multiFaMode && v.Chr == multiFaChromName { //only check if in multiFaMode and we have the right chrom
+			firstTime = false
+			prevPos = v.Pos
+		} else if multiFaMode {
+			if v.Pos <= prevPos && v.Chr == multiFaChromName {
+				log.Fatalf("Input VCF variants must be in sorted order in multiFaMode.")
+			}
+		}
+
 		if !(vcf.IsBiallelic(v) && vcf.IsSubstitution(v)) {
 			log.Fatal("Error: currently we only handle biallelic substitutions\n")
 		}
 
-		// check to make sure the fa matches the vcf before we edit
-		if seqsLookup[v.Chr][v.Pos-1] != dna.StringToBase(v.Ref) {
-			log.Fatal("Error: base in fasta didn't match ref base from VCF record\n")
-		}
+		if multiFaMode {
```

Comment:
I think what happened here is that there are two separate functions that have been combined into one function, but they only share about 4 lines of code and the other 100 lines are wrapped in "if" statements where we do one thing if we are in multiFa mode and another set of code if we are not.  One option that I think will be easier to read/review/maintain would be to add a separate function that does mutliFa mode.  This will not be duplicate code since the combined version does not really share code.
The other option would be to narrow the scope of this pull request to "improve" the code to allow gaps in the input sequence.
I think there are two things being added here: 1) we are allowing the reference sequence to have gaps in it.  The code that allows for gaps should also be fine for the case where we have no gaps, so I think we can just "improve" the current code rather than needing a new function 2) We are adding the ability to add a sequence to the original sequences in the file instead of needing to "cat" the input and output files.  In this case I think it would be better to just run "cat" on the input and output after running the program.  The workflow would be to faFilter the human sequence out of the multiFasta.  Run vcfToFa on the human fasta, which has gaps to edit it.  Then run "cat" on the multiFasta and the output to have the original alignment with the edited sequence added.
I worry that what could be a few commands are being put into one command, which could cause this command to be more difficult to a review/maintain.

Reply:
i went the first route in the next commit, split this function into two. However, both are still housed in the same cmd.